### ISSUE TITLE
Lazy loading autoplay videos

### ISSUE THREAD

**Issue by mondalaci** [aFarkas/lazysizes#697]
I'm trying to [lazy load autoplay videos](https://wordpress.org/support/topic/videos-dont-lazy-load/), but they get loaded without scrolling. Any ideas how to make them work?

**Comment by aFarkas** [comment]
I think I have to alter the markup here a little bit because you can use `autoplay=""` attribute. Maybe `data-autoplay=""`?

**Comment by mondalaci** [comment]
Yes, being able to use `data-autoplay=""` to lazy-load autoplay videos would be awesome!

**Comment by aFarkas** [comment]
@mondalaci 

I will only have time to do this after 10.10.19. In general it should be easy so you can give it a try.

**Comment by mondalaci** [comment]
Are you suggesting that I should implement this myself to test it? I would much rather wait for your implementation. It's not a pressing issue, and I better trust your code given your experience.

**Comment by rickydazla** [comment]
You can do this using the [unveilhooks extension](https://github.com/aFarkas/lazysizes/tree/master/plugins/unveilhooks) if you do something like this (or a data attribute), after it has changed the preload value to `auto`:

```
if(e.target.hasClass('autoplay') && e.target.autoplay != true){
  e.target.autoplay = true;
}
```

**Comment by aFarkas** [comment]
I just added the feature. You will need the following markup:

```html
<video
	class="lazyload"
	preload="none"
	muted=""
	data-autoplay=""
	data-poster="poster.jpg"
	src="src.mp4">
</video>
```

**Comment by rickydazla** [comment]
Love your work @aFarkas

**Comment by mondalaci** [comment]
@aFarkas The latest lazysizes version has been integrated into the beta version of the aforementioned WordPress plugin, and I can confirm that it's working as expected. Thank you so much for the great work!

**Comment by aerni** [comment]
@aFarkas This is only working halfway for me. Using the markup above the following happens:

**Chrome**
the video doesn't start playing. When I add `autoplay` it works. But this way the video gets loaded twice when looking at the network tab. I suppose `data-autoplay` has to be changed to `autoplay` after the content is loaded?

**Firefox**
The video gets lazyloaded, but the poster doesn't fade out and the video doesn't start playing. When adding `autoplay` it works as well. But same issue with double loading the video like in Chrome.

**Comment by aFarkas** [comment]
@aerni 
Could you please provide a testcase on jsfiddle, codepen, codesandbox or what ever.

**Comment by aerni** [comment]
@aFarkas Sure I can. To my clarification; Do you need the `unveilhooks plugin` to make this work or no plugin at all?

**Comment by aFarkas** [comment]
@aerni 
yes, this is only available via the unveilhooks plugin.

**Comment by willstocks** [comment]
@aFarkas - this appears to only work when you use `src=`, but not if/when you use multiple `<source>`'s - is that intentional?
Fortunately, in the case of the video I have on my site at the mo, I have a fairly well compressed `.mp4` so have swapped it into the `src` and it appears to work fine! But if I wanted `<source type="webm"><source type="mp4">` etc. this wouldn't work?

**Comment by willstocks** [comment]
Also, interestingly (and I haven't had a chance to debug yet), the solution does not appear to work on iOS 13.2 - the video just never loads for me on Safari/Chrome/Brave (obviously, all use the same underlying engine)

**Comment by aFarkas** [comment]
@aerni  and @willstocks 

I can't really reproduce your issues. Here is my testcase:
https://output.jsbin.com/qijoziq

**Comment by willstocks** [comment]
Thanks @afarkas - I will take a look/will spin up a test case when I get
home.

For the time being, I don’t know if this helps:
https://willstocks.co.uk/ scroll down to the “MPOW” article (last one of
four)

**Comment by aerni** [comment]
@aFarkas You're right. Your test case works as expected. I implemented your markup on my website but I still can't figure out why it's not working. The JS works as the class `lazyloaded` gets applied and the video lazyloaded. But the video just doesn't start playing. 

See here: [https://stage.michaelaerni.ch/en/work/hyve](https://stage.michaelaerni.ch/en/work/hyve)
Scroll down past the quote. There are two videos.

**Comment by aFarkas** [comment]
@aerni 
As it seems to me you don't have included the unveilhooks plugin.

**Comment by aerni** [comment]
@aFarkas I do have it included. It's bundled with Webpack in the `vendor.js`. 

```JS
import lazySizes from 'lazysizes';
import 'lazysizes/plugins/unveilhooks/ls.unveilhooks';
```

Funny enough; I just figured that it works fine when adding the `unveilhooks` plugin as an external script in my HTML like so: 

```HTML
<script src="https://afarkas.github.io/lazysizes/plugins/unveilhooks/ls.unveilhooks.min.js"></script>
```

Seems to be an issue with importing the plugin and bundling with Webpack?

**Comment by willstocks** [comment]
@aFarkas - I've just taken a look at that testcase and it appears to work as expected. I believe that, for some reason, regardless of my markup including `playsinline` and `-webkit-playsinline`, these two items are being stripped from my output for some reason... weird
Otherwise, lazyloading of the video works fine everywhere else!

**Comment by Stoyan0v** [comment]
> I can't really reproduce your issues. Here is my testcase:
> https://output.jsbin.com/qijoziq

Hey @aFarkas the testcase is not working. 
The lazyload is now working when the source is added in separate tag

**Comment by alexkuc** [comment]
Maybe this codepen will help someone - https://codepen.io/alex2407/pen/RwpJBJz

Some observations:

- it is `src`, **not** `data-src`
- it is `data-autoplay`, **not** `autoplay`
- `.container` is purely decorative to create large enough viewport to avoid loading video immediately
- `unveilhooks` plugin _is_ included

Initial page load:

![image](https://user-images.githubusercontent.com/3331946/121092542-6081d080-c7f4-11eb-87d9-5bd51800d06a.png)

Scrolling down until lazyloading happens:

![image](https://user-images.githubusercontent.com/3331946/121092598-7b544500-c7f4-11eb-8f7c-ce4add1aacb7.png)

Ps. I have used a free placeholder video, in case it breaks any copyright let me know and I will replace it with anything else; it is meant to be purely educational w/o any copyright infringement intentions

**Comment by onlinebizsoft** [comment]
> Maybe this codepen will help someone - https://codepen.io/alex2407/pen/RwpJBJz
> 
> Some observations:
> 
>     * it is `src`, **not** `data-src`
> 
>     * it is `data-autoplay`, **not** `autoplay`
> 
>     * `.container` is purely decorative to create large enough viewport to avoid loading video immediately
> 
>     * `unveilhooks` plugin _is_ included
> 
> 
> Initial page load:
> 
> ![image](https://user-images.githubusercontent.com/3331946/121092542-6081d080-c7f4-11eb-87d9-5bd51800d06a.png)
> 
> Scrolling down until lazyloading happens:
> 
> ![image](https://user-images.githubusercontent.com/3331946/121092598-7b544500-c7f4-11eb-8f7c-ce4add1aacb7.png)
> 
> Ps. I have used a free placeholder video, in case it breaks any copyright let me know and I will replace it with anything else; it is meant to be purely educational w/o any copyright infringement intentions

I have just tested and the webm seems to be loading once I open the page. In theory, when you use src then of course browser will load it once the HTML is parsed.

**Comment by alexkuc** [comment]
> I have just tested and the webm seems to be loading once I open the page

I have just tried and in my case, it opens when I scroll down to the video itself. I am on OSX with Google Chrome.

**Comment by evankford** [comment]
@alexkuc This works great on OSX, but doesn't work on iOS at all.

**Comment by willstocks** [comment]
Doesn't seem to be SUPER happy on iOS (any browser - Safari, Chrome, Brave - all open the video full screen even when playsinline is set)

**Comment by alexkuc** [comment]
@willstocks are you suggesting there is a bug or issue with my sample code? To be honest, my code is not meant to be used as a ready-copy-and-paste solution. But rather a guideline or reference to get started. If you can reproduce the bug, I think it's better to raise it as an actual bug issue.

**Comment by willstocks** [comment]
Hi @alexkuc - no, just lazyloaded video as a whole appears to still not be 100% happy autoplaying/playing inline on iOS. I did peek very quickly at your codepen, but I only see the poster image as I think webm video hasn't yet made its way to iOS (webm audio is though through a flag I think)

**Comment by gelsas** [comment]
I am completly new to this and I am trying to lazyload autoplay videos in my shopify stroe.

Can someone guide me in the right direction on what I need to do ?

**Comment by alexkuc** [comment]
@gelsas there is a "How To" section in [README.md](https://github.com/aFarkas/lazysizes/blob/gh-pages/README.md)

**Comment by alexchakmakov** [comment]
Hello everyone. I'm having the following markup

```
<video class="lazyload"
       preload="none"
       data-autoplay=""
       loop=""
       playsinline=""
>
    <source src="VIDEO_URL" type="video/mp4">
</video>
```

This setup works perfectly, as evidenced by the network tab showing that only 22 out of 60 videos load initially, with more videos loading incrementally as I scroll down (24, 26, 28, 30, etc.).

However, the problem arises when I add the muted attribute to the videos. When the videos have the muted attribute, it seems that lazysizes fails, and all 60 videos load immediately on page load. What could be causing this issue?


**Please note the following:**
- The videos are dynamically added to the DOM.
- I am using a masonry JavaScript library.
- I have an IntersectionObserver that plays the videos when they enter the viewport and pauses them when they exit.

Thanks in advance for you help! @aFarkas
